File: static/java/AirtableModel.java
// ==========================================
// MyAirtable static runtime.
// ==========================================

package myairtable;

import com.fasterxml.jackson.databind.JsonNode;
import java.time.Instant;
import java.util.Map;

/**
 * Implemented by every generated {@code {Table}Model} class. The generator emits the concrete class
 * with private {@code @JsonProperty} field-ID-keyed fields (computed fields getter-only, writable
 * fields with setters), a public no-arg constructor for Jackson, and a nested {@code Builder} over
 * the writable fields (Phase 0 gate, TestPojoModel.java).
 *
 * <p>Change detection for partial updates is handled with an explicit snapshot map captured after
 * decode/save — {@link #dirtyFields()} diffs against it.
 */
public interface AirtableModel {

  /** Table identity for the model. Generated models return their table's {@code TABLE_ID}. */
  String getTableId();

  /** Airtable record ID. {@code null} when the model has never been saved. Set by OrmTable. */
  String getId();

  void setId(String id);

  /** Server-assigned creation timestamp. {@code null} when unsaved. */
  Instant getCreatedTime();

  void setCreatedTime(Instant createdTime);

  /**
   * Airtable client attached by the table that produced this model. Set by {@link OrmTable} after a
   * successful decode so {@code save()}/{@code fetch()}/{@code delete()} can call back without the
   * caller threading a table through. Mirrors Rust's {@code ModelMeta.client} and Python's {@code
   * _at_client}.
   */
  AirtableClient getAttachedClient();

  void setAttachedClient(AirtableClient client);

  /** {@code true} when the model has no server-assigned ID yet. */
  default boolean isNew() {
    return getId() == null || getId().isEmpty();
  }

  /**
   * Capture the current writable field values as a snapshot. Called after decode and after a
   * successful save so subsequent {@link #dirtyFields()} computes the diff relative to server
   * state.
   */
  void takeSnapshot();

  /**
   * Writable fields that changed since the last snapshot, keyed by field ID. Cleared fields appear
   * as JSON null. Used by {@link OrmTable#update} to send only what changed.
   */
  Map<String, JsonNode> dirtyFields();

  /** All current non-null field values, keyed by field ID. */
  Map<String, JsonNode> toRecord();

  /**
   * Non-null <b>writable</b> field values, keyed by field ID — the only path to create payloads.
   * Computed fields never appear here.
   */
  Map<String, JsonNode> toCreateFields();

  /**
   * The server-assigned record ID, or an UNSAVED_MODEL error.
   *
   * <p>Every model returned by a table's {@code create(...)} / {@code get(...)} carries an ID, so
   * prefer this over a manual null check — a {@code null} ID raises a descriptive {@link
   * AirtableException.Api} instead of a bare {@link NullPointerException}.
   */
  default String requireId() {
    String id = getId();
    if (id == null || id.isEmpty()) {
      throw new AirtableException.Api("UNSAVED_MODEL", "Model has no server-assigned id yet");
    }
    return id;
  }

  /** The attached client, or a DETACHED_MODEL error. */
  default AirtableClient requireAttachedClient() {
    AirtableClient client = getAttachedClient();
    if (client == null) {
      throw new AirtableException.Api(
          "DETACHED_MODEL",
          "Model must be obtained via a table (get / create / upsert) before calling save() /"
              + " fetch() / delete()");
    }
    return client;
  }
}

File: static/csharp/AirtableModel.cs
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;

namespace MyAirtable;

/// <summary>
/// Abstract base for every generated <c>{Table}Model</c> class. Holds the record identity
/// (<see cref="Id"/> / <see cref="CreatedTime"/>), the <see cref="AttachedClient"/> wired in by
/// the producing <see cref="OrmTable{T}"/>, and an explicit writable-field snapshot used for
/// partial-update change detection.
///
/// <para>Generated subclasses contribute their field values through the two
/// <c>Collect…</c> overrides; everything else — snapshotting, dirty diffing, payload building —
/// lives here so the generator emits no boilerplate (csharp-plan-v2 §2.3.1). STJ decodes the
/// typed auto-properties wholesale; the payload maps below are hand-rolled writable-only.</para>
/// </summary>
public abstract class AirtableModel
{
    private static readonly IReadOnlyDictionary<string, JsonNode?> EmptySnapshot =
        new Dictionary<string, JsonNode?>();

    /// <summary>Table identity. Generated models return their table's <c>TableId</c> constant.</summary>
    [JsonIgnore]
    public abstract string TableId { get; }

    /// <summary>Airtable record ID. <c>null</c> until the model has been saved; set by the table.</summary>
    [JsonIgnore]
    public string? Id { get; set; }

    /// <summary>Server-assigned creation timestamp. <c>null</c> when unsaved.</summary>
    [JsonIgnore]
    public DateTimeOffset? CreatedTime { get; set; }

    /// <summary>
    /// Client attached by the table that produced this model (after a successful decode), so the
    /// fluent <c>SaveAsync</c>/<c>FetchAsync</c>/<c>DeleteAsync</c> can call back without the caller
    /// threading a table through. Mirrors Rust's <c>ModelMeta.client</c>.
    /// </summary>
    [JsonIgnore]
    public AirtableClient? AttachedClient { get; set; }

    [JsonIgnore]
    private Dictionary<string, JsonNode?>? _snapshot;

    /// <summary><c>true</c> when the model has no server-assigned ID yet.</summary>
    [JsonIgnore]
    public bool IsNew => string.IsNullOrEmpty(Id);

    /// <summary>
    /// Current writable field values keyed by field ID. EVERY writable field appears (with a
    /// <c>null</c> value when unset) so <see cref="DirtyFields"/> can detect clears. Implemented by
    /// the generated subclass.
    /// </summary>
    protected abstract IReadOnlyDictionary<string, JsonNode?> CollectWritableFields();

    /// <summary>Current computed (read-only) field values keyed by field ID. Implemented by the generated subclass.</summary>
    protected abstract IReadOnlyDictionary<string, JsonNode?> CollectComputedFields();

    /// <summary>
    /// Capture the current writable field values so subsequent <see cref="DirtyFields"/> diffs
    /// against this point. The table snapshots on every decode, so a table-produced model starts clean.
    ///
    /// <para>Note: a save returns a FRESH, freshly-snapshotted instance; the model you called save on
    /// keeps its pre-save snapshot. Use the returned instance for further edits.</para>
    /// </summary>
    public void TakeSnapshot()
    {
        var snapshot = new Dictionary<string, JsonNode?>();
        foreach (var (id, value) in CollectWritableFields())
            snapshot[id] = value?.DeepClone();
        _snapshot = snapshot;
    }

    /// <summary>
    /// Writable fields that changed since the last <see cref="TakeSnapshot"/>, keyed by field ID.
    /// Cleared fields appear as a <c>null</c> value. Used by <see cref="OrmTable{T}"/> to PATCH only
    /// what changed.
    /// </summary>
    public Dictionary<string, JsonNode?> DirtyFields()
    {
        var snapshot = _snapshot ?? EmptySnapshot;
        var dirty = new Dictionary<string, JsonNode?>();
        foreach (var (id, value) in CollectWritableFields())
        {
            snapshot.TryGetValue(id, out var previous);
            if (!AirtableRuntime.IsEqual(previous, value))
                dirty[id] = value;
        }
        return dirty;
    }

    /// <summary>All current non-null field values (writable + computed), keyed by field ID.</summary>
    public Dictionary<string, JsonNode?> ToRecord()
    {
        var record = new Dictionary<string, JsonNode?>();
        foreach (var (id, value) in CollectWritableFields())
            if (value is not null)
                record[id] = value;
        foreach (var (id, value) in CollectComputedFields())
            if (value is not null)
                record[id] = value;
        return record;
    }

    /// <summary>
    /// Non-null <b>writable</b> field values keyed by field ID — the only path to create payloads.
    /// Computed fields never appear here.
    /// </summary>
    public Dictionary<string, JsonNode?> ToCreateFields()
    {
        var fields = new Dictionary<string, JsonNode?>();
        foreach (var (id, value) in CollectWritableFields())
            if (value is not null)
                fields[id] = value;
        return fields;
    }

    /// <summary>
    /// The server-assigned record ID, or an <see cref="AirtableException.ApiError"/> with code
    /// <c>UNSAVED_MODEL</c>. Every model returned by a table carries an ID, so prefer this over a
    /// manual null check.
    /// </summary>
    public string RequireId() =>
        IsNew
            ? throw new AirtableException.ApiError(
                "UNSAVED_MODEL",
                "Model has no server-assigned id yet"
            )
            : Id!;

    /// <summary>The attached client, or an <see cref="AirtableException.ApiError"/> with code <c>DETACHED_MODEL</c>.</summary>
    public AirtableClient RequireAttachedClient() =>
        AttachedClient
        ?? throw new AirtableException.ApiError(
            "DETACHED_MODEL",
            "Model must be obtained via a table (Get / Create / Upsert) before calling "
                + "SaveAsync / FetchAsync / DeleteAsync"
        );
}
